src/bridge/crypto/ec.cc

```diff
@@ -0,0 +1,303 @@
+  StatusOr<CryptoKeyHandle *> crypto_key_handle_or =
+      GetCryptoKeyHandleFromOpenSslEcKey(ec_key);
+  if (crypto_key_handle_or.ok()) {
+    delete crypto_key_handle_or.value();
+  }
+
+  Status status = AttachCryptoKeyHandleToOpenSslEcKey(nullptr, ec_key);
+  if (!status.ok()) {
+    KMSENGINE_SIGNAL_ERROR(status);
+  }
+}
+
+// Called at the end of `EC_KEY_copy`, which copies the contents of `src` into
+// `dest`.
+//
+// Returns 1 on success; otherwise, returns 0.
+int Copy(EC_KEY *dest, const EC_KEY *src) {
+  // `crypto_key_handle` is guaranteed to be non-null here (if the underlying
+  // external data struct was null, an error status would be returned).
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      const CryptoKeyHandle *src_handle,
+      GetCryptoKeyHandleFromOpenSslEcKey(src), 0);
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      std::unique_ptr<CryptoKeyHandle> dest_handle,
+      backing::CopyCryptoKeyHandle(*src_handle), 0);
+
+  Status status = AttachCryptoKeyHandleToOpenSslEcKey(
+      std::move(dest_handle), dest);
+  if (!status.ok()) {
+    KMSENGINE_SIGNAL_ERROR(status);
+    return 0;
+  }
+  return 1;
+}
+
+// Engine implementation for `ECDSA_sign_ex`.
+//
+// Computes a digital signature of the hash value `digest_bytes` of length
+// `digest_length` using the private EC key `ec_key`.
+//
+// According to ECDSA_sign_ex(3), the parameter `type` is ignored. Additionally,
+// `kinv` and `rp` are optional parameters that are used to precompute parts of
+// the signing operation (we ignore those here since the engine is not
+// performing any crypto calculations itself).
+//
+// Returns 1 on success; otherwise, returns 0.
+int SignEx(int type, const unsigned char *digest_bytes, int digest_length,
+           unsigned char *signature_return, unsigned int *signature_length,
+           const BIGNUM */*kinv*/, const BIGNUM */*rp*/, EC_KEY *ec_key) {
+  // Convert arguments to engine-native structures for convenience. These
+  // conversions need to take place within the bridge layer (as opposed to
+  // letting the `RsaKey::Sign` method handling the conversions) since the
+  // conversion functions refer to some OpenSSL API functions.
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      const CryptoKeyHandle *crypto_key_handle,
+      GetCryptoKeyHandleFromOpenSslEcKey(ec_key), 0);
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      const DigestCase digest_type,
+      ConvertOpenSslNidToDigestType(type), 0);
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      const std::string digest_string,
+      MakeDigestString(digest_bytes, digest_length), 0);
+
+  // Delegate handling of the signing operation to the backing layer.
+  KMSENGINE_ASSIGN_OR_RETURN_WITH_OPENSSL_ERROR(
+      const std::string signature,
+      crypto_key_handle->Sign(digest_type, digest_string), 0);
+
+  // Copy results into the return pointers.
+  if (signature_length != nullptr) {
```

Comment:
If we want to be consistent with RSA signing, then raising an error when `signature_length == nullptr` seems to make sense.

Reply:
For future work---see https://github.com/googleinterns/cloud-kms-oss-tools/pull/114#discussion_r469477447, which outlines the error-checking strategy on `signature_length` and `signature_return` in RSA signing.

Actually, the code for both EC's `SignEx` and RSA's `Sign` are identical except for the `GetCryptoKeyHandleFromOpenSslRsa` vs `GetCryptoKeyHandleFromOpenSslEcKey` prior to this. One could define a `CommonSign` function that has a signature of:

```c++
Status CommonSign(int type, const unsigned char *digest_bytes, int digest_length,
                  unsigned char *signature_return, unsigned int *signature_length, CryptoKeyHandle *handle);
```

And just redirect `SignEx` and RSA's `Sign` to `CommonSign`, or something similar.